subsys/bluetooth/audio/has_client.c

```diff
@@ -82,11 +88,195 @@ static enum bt_has_capabilities get_capabilities(const struct has_inst *inst)
+	};
+};
+
+static void handle_read_preset_rsp(struct bt_conn *conn, struct net_buf_simple *buf)
+{
+	struct has_inst *inst = inst_by_conn(conn);
+	const struct bt_has_cp_read_preset_rsp *rsp;
+	union preset_record record;
+	size_t name_len;
+
+	BT_DBG("conn %p buf %p", (void *)conn, buf);
+
+	if (buf->len < sizeof(*rsp)) {
+		BT_ERR("malformed PDU");
+		return;
+	}
+
+	rsp = net_buf_simple_pull_mem(buf, sizeof(*rsp));
+
+	record.properties = rsp->properties;
+	name_len = MIN(buf->len + 1, ARRAY_SIZE(record.name));
+	utf8_lcpy(record.name, rsp->name, name_len);
+
+	if (rsp->is_last > 0x01) {
+		BT_WARN("unexpected is_last 0x%02x", rsp->is_last);
+	}
+
+	client_cb->preset_event(&inst->has, 0, BT_HAS_PRESET_RECORD_READ_RSP, rsp->index,
+				&record.record, !!rsp->is_last);
+}
+
+static void handle_generic_update(struct bt_conn *conn, struct net_buf_simple *buf, bool is_last)
+{
+	struct has_inst *inst;
+	const struct bt_has_cp_generic_update *generic_update;
+	struct bt_has_preset_delete delete;
+	union preset_record record;
+	size_t name_len;
+
+	if (buf->len < sizeof(*generic_update)) {
+		BT_ERR("malformed PDU");
+		return;
+	}
+
+	inst = inst_by_conn(conn);
+
+	generic_update = net_buf_simple_pull_mem(buf, sizeof(*generic_update));
+
+	/* Generic Update can be used as well to notify preset deletion.
+	 * Notify the user that there are no presets registered in range (PrevIndex, Index)
+	 */
+	delete.end_index = generic_update->index - 1;
+	client_cb->preset_event(&inst->has, 0, BT_HAS_PRESET_DELETE,
+				generic_update->prev_index + 1, &delete, false);
+
+	record.properties = generic_update->properties;
+	name_len = MIN(buf->len + 1, ARRAY_SIZE(record.name));
+	utf8_lcpy(record.name, generic_update->name, name_len);
+
+	client_cb->preset_event(&inst->has, 0, BT_HAS_PRESET_RECORD_UPDATE,
+				generic_update->index, &record.record, is_last);
+}
+
+static void handle_preset_deleted(struct bt_conn *conn, struct net_buf_simple *buf, bool is_last)
+{
+	struct has_inst *inst;
+	struct bt_has_preset_delete delete;
+	uint8_t index;
+
+	if (buf->len < sizeof(index)) {
+		BT_ERR("malformed PDU");
+		return;
+	}
+
+	inst = inst_by_conn(conn);
+
+	index = net_buf_simple_pull_u8(buf);
+
+	delete.end_index = index;
+
+	client_cb->preset_event(&inst->has, 0, BT_HAS_PRESET_DELETE, index, &delete, is_last);
+}
+
+static void handle_preset_available(struct bt_conn *conn, struct net_buf_simple *buf, bool is_last)
+{
+	struct has_inst *inst;
+
+	if (buf->len < sizeof(uint8_t)) {
+		BT_ERR("malformed PDU");
+		return;
+	}
+
+	inst = inst_by_conn(conn);
+
+	client_cb->preset_event(&inst->has, 0, BT_HAS_PRESET_AVAILABLE,
+				net_buf_simple_pull_u8(buf), NULL, is_last);
+}
+
+static void handle_preset_unavailable(struct bt_conn *conn, struct net_buf_simple *buf,
+				      bool is_last)
+{
+	struct has_inst *inst;
+
+	if (buf->len < sizeof(uint8_t)) {
+		BT_ERR("malformed PDU");
+		return;
+	}
+
+	inst = inst_by_conn(conn);
```

Comment:
Instead of doing `inst = inst_by_conn(conn);` in each handler function, we should do that in the caller and provide a `struct has_inst *inst` instead of `struct bt_conn *conn`